### PR TITLE
[WIP] Add TorchScript compatibility for LightningModules

### LINKED ISSUE #1951: Support torch.jit.script on LightningModules
## 🚀 Feature

There are a number of advantages to converting a model with TorchScript (e.g. static optimizations, better saving / loading, especially into non-Python environments for deployment). However, no `LightningModule`s can be converted using `torch.jit.script`. Here's a simple example with the error produced (note that this works as-is if we inherit from `nn.Module` instead of `pl.LightningModule`):

```python
import pytorch_lightning as pl
import torch

# class Model(nn.Module): # works fine
class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(5, 10)
    
    def forward(self, x):
        return self.layer(x)

torch.jit.script(Model())
```

```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-70-1fe19c1470da> in <module>
     10         return self.layer(x)
     11 
---> 12 torch.jit.script(Model())

~/anaconda/envs/vsrl/lib/python3.7/site-packages/torch/jit/__init__.py in script(obj, optimize, _frames_up, _rcb)
   1259 
   1260     if isinstance(obj, torch.nn.Module):
-> 1261         return torch.jit._recursive.create_script_module(obj, torch.jit._recursive.infer_methods_to_compile)
   1262 
   1263     qualified_name = _qualified_name(obj)

~/anaconda/envs/vsrl/lib/python3.7/site-packages/torch/jit/_recursive.py in create_script_module(nn_module, stubs_fn, share_types)
    295     if share_types:
    296         # Look into the store of cached JIT types
--> 297         concrete_type = concrete_type_store.get_or_create_concrete_type(nn_module)
    298     else:
    299         # Get a concrete type directly, without trying to re-use an existing JIT

~/anaconda/envs/vsrl/lib/python3.7/site-packages/torch/jit/_recursive.py in get_or_create_concrete_type(self, nn_module)
    254             return nn_module._concrete_type
    255 
--> 256         concrete_type_builder = infer_concrete_type_builder(nn_module)
    257 
    258         nn_module_type = type(nn_module)

~/anaconda/envs/vsrl/lib/python3.7/site-packages/torch/jit/_recursive.py in infer_concrete_type_builder(nn_module)
    133     # Constants annotated via `Final[T]` rather than being added to `__constants__`
    134     for name, ann in class_annotations.items():
--> 135         if torch._jit_internal.is_final(ann):
    136             constants_set.add(name)
    137 

~/anaconda/envs/vsrl/lib/python3.7/site-packages/torch/_jit_internal.py in is_final(ann)
    681 
    682     def is_final(ann):
--> 683         return ann.__module__ == 'typing_extensions' and \
    684             (getattr(ann, '__origin__', None) is typing_extensions.Final)
    685 except ImportError:

AttributeError: 'ellipsis' object has no attribute '__module__'
```

Digging into this a little, we have

```python
print(Model.__annotations__)
# {'_device': Ellipsis, '_dtype': typing.Union[str, torch.dtype]}
```

and the `_device` annotation comes from [`DeviceDtypeModuleMixin`] (one of the super-classes of `LightningModule`). Here's the relevant snippet:

```python
class DeviceDtypeModuleMixin(torch.nn.Module):
    _device: ...
```

This seems to be the only issue because this code works:

```python
import pytorch_lightning as pl
import torch

class Model(pl.LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(5, 10)
    
    def forward(self, x):
        return self.layer(x)

# Model.__annotations__ = {} # this works too but doesn't seem as nice
Model.__annotations__["_device"] = torch.device
torch.jit.script(Model())
```

However, if I try to set the annotation to `typing.Union[str, torch.device]` (which seems to be the true type based on [this line]), then I get `ValueError: Unknown type annotation: 'typing.Union[str, torch.device]' in `torch.jit.script`.

Is the `str` type for `_device` actually used? I don't see that anywhere, and I actually do see at least one place where there would be an error if `self.device` returned a string ([here]. I'll just go ahead and submit a PR to update the annotations, but feel free to comment here or on the PR if there's something I'm missing about the type annotations here.

[`DeviceDtypeModuleMixin`]: https://github.com/PyTorchLightning/pytorch-lightning/blob/d0ec11b9d656a185b10a113a08dfba3042b4da01/pytorch_lightning/utilities/device_dtype_mixin.py#L6
[this line]: https://github.com/PyTorchLightning/pytorch-lightning/blob/d0ec11b9d656a185b10a113a08dfba3042b4da01/pytorch_lightning/utilities/device_dtype_mixin.py#L20
[here]: https://github.com/PyTorchLightning/pytorch-lightning/blob/d0ec11b9d656a185b10a113a08dfba3042b4da01/pytorch_lightning/core/lightning.py#L88

### PR DESCRIPTION
**Pull request by neighthan**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
  - The issue is [here] (but there's been no discussion / approval)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
  - Didn't seem necessary
- [ ] Did you write any new necessary tests?
  - This should probably be done if we want TorchScript to work going forward; if that's not a priority, then no tests are needed (this fix could be merged without necessarily committing to long-term TorchScript support)
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)?
  - Didn't seem necessary

## What does this PR do?
Fixes #1951.

[here]: https://github.com/PyTorchLightning/pytorch-lightning/issues/1951

### CODE REVIEW

**Comment by neighthan** [comment]
I didn't test this enough; I think this isn't complete yet.

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 1]
```diff
@@ -4,11 +4,11 @@
 
 
 class DeviceDtypeModuleMixin(torch.nn.Module):
-    _device: ...
-    _dtype: Union[str, torch.dtype]
+    _device: torch.device
+    _dtype: torch.dtype
```
Comment (justusschock) [inline pytorch_lightning/utilities/device_dtype_mixin.py:8]: why shouldn't I be omit strings here?

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 2]
```diff
@@ -4,11 +4,11 @@
...
 
     @property
-    def dtype(self) -> Union[str, torch.dtype]:
+    def dtype(self) -> torch.dtype:
```
Comment (justusschock) [inline pytorch_lightning/utilities/device_dtype_mixin.py:11]: Same for strings

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 3]
```diff
@@ -17,7 +17,7 @@ def dtype(self, new_dtype: Union[str, torch.dtype]):
         raise RuntimeError('Cannot set the dtype explicitly. Please use module.to(new_dtype).')
 
     @property
-    def device(self) -> Union[str, torch.device]:
+    def device(self) -> torch.device:
```
Comment (justusschock) [inline pytorch_lightning/utilities/device_dtype_mixin.py:20]: Again strings :D

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 4]
```diff
@@ -125,6 +125,8 @@ def type(self, dst_type: Union[str, torch.dtype]) -> torch.nn.Module:
         Returns:
             Module: self
         """
+        if isinstance(dst_type, str):
```
Comment (justusschock) [inline pytorch_lightning/utilities/device_dtype_mixin.py:128]: not sure this is working for all the types. Even if it is. Can't we maybe find a better way to do this? For devices, you can simply do `device = torch.device(device_str)`. I know, this does not work for dtypes, but maybe something similar? But in general: why doesn't strings work with torchscript?

**Comment by awaelchli** [comment]
It would be good to have a test that torch script can compile the template we use in tests, because the device properties were added only recently and it seems that it broke the compatibility.

**Comment by Borda** [comment]
@neighthan how is it going here? :rabbit:

**Comment by neighthan** [comment]
Could you add the help-wanted label here? The project where I was hoping to use TorchScript didn't benefit from it, so writing the tests that should be in this PR / checking the issues with `str` dtypes is low-priority for me right now.

**Comment by lezwon** [comment]
@awaelchli @Borda I'd like to work on this issue. Would require a little help on what would be a good approach to fix this though :]

**Comment by justusschock** [comment]
@lezwon It's basically adding the missing tests (like you did with onnx) and then checking the dtype-str issue. The rest should be fine from what I saw :)

**Comment by lezwon** [comment]
@justusschock cool :) I'll get started with it then. 👍

**Comment by neighthan** [comment]
@lezwon this is resolved in your PR [here](https://github.com/PyTorchLightning/pytorch-lightning/pull/2657/files), right? I'll close this now

**Pull request closed without merging**